### PR TITLE
feat: use react-i18next for translation

### PR DESCRIPTION
**Pull request by WindSpiritSR**
fixed: #168 

Signed-off-by: WindSpiritSR <simon343riley@gmail.com>

### CODE REVIEW

**Comment by ErikQQY** [comment]
Wow, it’s awesome, maybe we should communicate with each other in advance. That issue is an idea, I think that feature could bring comfortable experiences to our users, do you have demo links for this PR? I would love to collaborate with you on this feature.

**Comment by WindSpiritSR** [comment]
This is a great idea and easy to implement.

Deployed at https://door.leviatan.cn

**Comment by hsluoyz** [comment]
@WindSpiritSR only "Home" is translated:

image

**Comment by WindSpiritSR** [comment]
@hsluoyz `locales/zh/data.json` doesn't translate

**Comment by ErikQQY** [comment]
I think it is because since [this commit](https://github.com/casbin/casdoor/commit/2297251dd7523892983ce46d3289a45f597828be) the translations going wrong and there are no translations, I am checking on it

File: web/src/Setting.js [hunk 1]
```diff
@@ -285,18 +285,25 @@ export function getAvatarColor(s) {
   return colorList[random % 4];
 }
 
+export const I18n = (text) => {
+  return (
+    <Translation>
```
Comment (sh1luo) [inline web/src/Setting.js:290]: Why not continue to use `i18n`, `i` is lowercase.

**Review by sh1luo — COMMENTED** [review]
Code looks good, but I want to know what is the effect after replacement using `react-i18next`, is the performance better, or other effects?

**Comment by ErikQQY** [comment]
Sorry for the Crowdin commit mistake, I think it is all good now

casdoortrans

Next time a commit is pushed to the repo, the translation will working

@WindSpiritSR If you want to try the Chinese translation on you demo, you can substitute your original Chinese translation to [this](https://github.com/ErikQQY/casdoor/blob/master/web/src/locales/zh/data.json), just use this data.json instead of the original in your repo.

**Comment by ErikQQY** [comment]
> Code looks good, but I want to know what is the effect after replacement using `react-i18next`, is the performance better, or other effects?

After clicking other languages, the pages will not refresh, but directly change to the language you choose

**Comment by hsluoyz** [comment]
@ErikQQY the translation is still wrong, also:

1. Put the language selection to top-right, a select box.
2. Show an antd alert to know language is changed. The current way is too silent.

**Review by hsluoyz — CHANGES_REQUESTED** [review]
@WindSpiritSR @ErikQQY the translation is still wrong, also:

1. Put the language selection to top-right, a select box.
2. Show an antd alert to know language is changed. The current way is too silent.

**Comment by WindSpiritSR** [comment]
```js
export function I18n(text) {
  return (
    <Translation>
      {t => {return t(text)}}
    </Translation>
  )
}
```

These thing was **wrong**, because this function return a DOM object, if the text like `<input placeholder={Setting.I18n("text")}/>`, we will get a input box that shows `[object object]`

The right way to use `react-i18next`:

```js
import { withTranslation } from 'react-i18next';
@withTranslation()
class Test extends Component {
  //...code
  {this.props.t("Text")}
  //...code
}
export default Test;
```

or

```js
import { withTranslation } from 'react-i18next';
class Test extends Component {
  //...code
  {this.props.t("Text")}
  //...code
}
export default withTranslation()(Test);
```

This usage needs a `class`, but sometimes we need to translate at a `function`, I have no ideas to make it usable and look good. :-(

File: web/src/Setting.js (hunk 1 above)
Comment (hsluoyz) [reply web/src/Setting.js:290]: @sh1luo I agree

@WindSpiritSR plz change

**Comment by hsluoyz** [comment]
@WindSpiritSR yeah, I also see this `[object object]` bug.

> These thing was wrong, because this function return a DOM object, if the text like <input placeholder={Setting.I18n("text")}/>, we will get a input box that shows [object object]

I don't understand, this is your PR, right? Why do you say this is wrong instead of fixing it? It's just like you are reviewing a PR from others.

**Comment by hsluoyz** [comment]
@WindSpiritSR and plz fix:

image

**Pull request closed without merging**